### PR TITLE
Implement bridge

### PR DESCRIPTION
**Pull request by xepozz**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️/❌
| New feature?  | ✔️/❌
| Breaks BC?    | ✔️/❌
| Fixed issues  | comma-separated list of tickets # fixed by the PR, if any

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -19,7 +19,10 @@
     "minimum-stability": "dev",
     "prefer-stable": true,
     "require": {
-        "php": "^7.4|^8.0"
+        "php": "^7.4|^8.0",
+        "guzzlehttp/guzzle": "^7.3",
```
Comment (samdark) [inline composer.json]: I'm not sure it is a good idea to enforce usage of Guzzle. How about using PSR-18 instead?

File: composer.json [hunk 2]
```diff
@@ -40,6 +43,10 @@
     "extra": {
         "branch-alias": {
             "dev-master": "3.0.x-dev"
+        },
+        "config-plugin": {
+            "params": "config/params.php",
+            "common": "config/common.php"
```
Comment (vjik) [inline composer.json]:
```suggestion
        "config-plugin-options": {
            "source-directory": "config" 
        },
        "config-plugin": {
            "params": "params.php",
            "common": "common.php"
```

File: config/providers.php [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+declare(strict_types=1);
+
+use Yiisoft\Yii\Sentry\Provider\SentryProvider;
+
+return [
```
Comment (rustamwin) [inline config/providers.php:17]: What about ability to disable provider in params?

File: composer.json (hunk 1 above)
Comment (xepozz) [reply composer.json]: Agreed, but
```
$ composer req sentry/sentry
Using version ^3.3 for sentry/sentry
./composer.json has been updated
Running composer update sentry/sentry
Loading composer repositories with package information
Updating dependencies
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - sentry/sentry[dev-master, 3.3.0] require php-http/async-client-implementation ^1.0 -> could not be found in any version, but the following packages provide it:
      - symfony/http-client Provides powerful methods to fetch HTTP resources synchronously or asynchronously
      - php-http/guzzle6-adapter Guzzle 6 HTTP Adapter
      - symfony/symfony The Symfony PHP framework
      - php-http/curl-client PSR-18 and HTTPlug Async client with cURL
      - php-http/guzzle7-adapter Guzzle 7 HTTP Adapter
      - php-http/mock-client Mock HTTP client
      - voku/httpful A Readable, Chainable, REST friendly, PHP HTTP Client
      - php-http/react-adapter React HTTP Adapter
      - swisnl/php-http-fixture-client Fixture client for PHP-HTTP
      - phpgt/fetch Asynchronous HTTP client with promises for PHP 7 applications.
      - betalabs/guzzle6-adapter Guzzle 6 HTTP Adapter
      - neur0toxine/pock PSR-18 compatible HTTP mock library
      - nessworthy/amphp-http-httplug A small crappy adapter package to support php-http/async-client-implementation via. amphp/http-clien
      - phpgt/curl cURL object wrapper.
      - john_zuk/php-nbp-api NBP (Narodowy Bank Polski) API client
      - christophwurst/nextcloud-http-client HTTPlug and PSR7 adapter for Nextcloud
      - devster/http-client Symfony HttpClient component
      - gogoprint/guzzle6-adapter Guzzle 6 HTTP Adapter
      - pbxapi/php-http-fixture-client Fixture client for PHP-HTTP
      - phpgt/curlinterface cURL object wrapper.
      ... and 2 more.
      Consider requiring one of these to satisfy the php-http/async-client-implementation requirement.
    - sentry/sentry 3.3.x-dev is an alias of sentry/sentry dev-master and thus requires it to be installed too.
    - Root composer.json requires sentry/sentry ^3.3 -> satisfiable by sentry/sentry[3.3.0, 3.3.x-dev (alias of dev-master)].

Potential causes:
 - A typo in the package name
 - The package is not available in a stable-enough version according to your minimum-stability setting
   see <https://getcomposer.org/doc/04-schema.md#minimum-stability> for more details.
 - It's a private package and you forgot to add a custom repository to find it

Read <https://getcomposer.org/doc/articles/troubleshooting.md> for further common problems.

Installation failed, reverting ./composer.json and ./composer.lock to their original content.
```

Comment (samdark) [reply composer.json]: Well, you can add Guzzle as a dev dependency, but not as production one.

File: src/Middleware/SentryIntegration.php [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+namespace Yiisoft\Yii\Sentry\Middleware;
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Sentry\State\HubInterface;
+use Throwable;
+
+final class SentryIntegration implements MiddlewareInterface
```
Comment (samdark) [inline src/Middleware/SentryIntegration.php]: MiddlewareInterface requires `psr/http-server-middleware` in composer's `require`.

Comment (xepozz) [reply src/Middleware/SentryIntegration.php]: Done

File: composer.json (hunk 1 above)
Comment (xepozz) [reply composer.json]: Done

**Comment by samdark** [comment]
@xepozz is it still work in progress or ready for review?

**Comment by xepozz** [comment]
It's ready. Another features will be added later.

File: composer.json [hunk 5]
```diff
@@ -19,9 +19,16 @@
     "minimum-stability": "dev",
     "prefer-stable": true,
     "require": {
-        "php": "^7.4|^8.0"
+        "php": "^7.4|^8.0",
+        "php-http/httplug": "^2.2",
```
Comment (samdark) [inline composer.json]: I think it worth moving this dependency to require-dev and requiring https://packagist.org/providers/psr/http-client-implementation instead.

File: composer.json [hunk 6]
```diff
@@ -19,9 +19,16 @@
...
+        "psr/http-message": "^1.0",
+        "psr/http-server-middleware": "^1.0",
+        "sentry/sentry": "^3.3",
+        "yiisoft/di": "3.0.x-dev"
     },
     "require-dev": {
+        "guzzlehttp/guzzle": "^7.3",
```
Comment (samdark) [inline composer.json:29]: Why getting both guzzle and http plug at the same time?

File: src/Middleware/SentryIntegration.php (hunk 4 above)
Comment (samdark) [inline src/Middleware/SentryIntegration.php]:
```suggestion
final class SentryMiddleware implements MiddlewareInterface
```

**Comment by samdark** [comment]
Since there are only two classes, I think we can put them to the package root namespace without sub-namespaces.

File: composer.json (hunk 5 above)
Comment (xepozz) [reply composer.json]: It's impossible. Sentry needs async http client. 
See https://github.com/yiisoft/yii-sentry/pull/1#discussion_r650594020

File: composer.json (hunk 6 above)
Comment (xepozz) [reply composer.json:29]: httpplug needs to configure abstractions in config files. [httplug doesn't provide an adapter](https://github.com/php-http/httplug/tree/master/src).

And yes, I need to move `guzzle` to `require` section too. Are you agree with it or not?

File: src/SentryProvider.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+use Sentry\State\HubInterface;
+use Sentry\Transport\TransportFactoryInterface;
+use Yiisoft\Di\Container;
+use Yiisoft\Di\Support\ServiceProvider;
+
+final class SentryProvider extends ServiceProvider
+{
+    /**
+     * @psalm-suppress InaccessibleMethod
+     */
```
Comment (vjik) [inline src/SentryProvider.php]:
```suggestion
```

no need

File: composer.json (hunk 6 above)
Comment (samdark) [reply composer.json:29]: I'd prefer not to tie to concrete implementation.

File: composer.json (hunk 5 above)
Comment (samdark) [reply composer.json]: Can't we depend on [HTTP client PSR](https://www.php-fig.org/psr/psr-18/)? `psr/http-client-implementation`

**Comment by samdark** [comment]
@xepozz let's require defining your own PSR implementation and use interface instead.

**Comment by xepozz** [comment]
I moved packages to dev requirements and moved configuration into userland.
See https://github.com/yiisoft/app-api/pull/85 if you want to try how it works.

File: config/common.php [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+/**
+ * @var $params array
+ */
+
+return [
+    \Sentry\Transport\TransportFactoryInterface::class => \Sentry\Transport\DefaultTransportFactory::class,
+    \Sentry\HttpClient\HttpClientFactoryInterface::class => function (Yiisoft\Injector\Injector $injector) {
+        return $injector->make(\Sentry\HttpClient\HttpClientFactory::class, [
+            'sdkIdentifier' => \Sentry\Client::SDK_IDENTIFIER,
+            // TODO use composer tool
```
Comment (samdark) [inline config/common.php]: ?

File: config/common.php [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+declare(strict_types=1);
+
+/**
+ * @var $params array
+ */
+
+return [
+    \Sentry\Transport\TransportFactoryInterface::class => \Sentry\Transport\DefaultTransportFactory::class,
+    \Sentry\HttpClient\HttpClientFactoryInterface::class => function (Yiisoft\Injector\Injector $injector) {
+        return $injector->make(\Sentry\HttpClient\HttpClientFactory::class, [
```
Comment (samdark) [inline config/common.php]: I think we can safely do `new` here instead of using injector.

File: config/common.php [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+
+declare(strict_types=1);
+
+/**
+ * @var $params array
+ */
+
+return [
+    \Sentry\Transport\TransportFactoryInterface::class => \Sentry\Transport\DefaultTransportFactory::class,
+    \Sentry\HttpClient\HttpClientFactoryInterface::class => function (Yiisoft\Injector\Injector $injector) {
```
Comment (samdark) [inline config/common.php]:
```suggestion
    \Sentry\HttpClient\HttpClientFactoryInterface::class => static function (Yiisoft\Injector\Injector $injector) {
```

File: composer.json (hunk 5 above)
Comment (xepozz) [reply composer.json]: No

File: config/common.php (hunk 9 above)
Comment (xepozz) [reply config/common.php]: Not really, but I rewrote syntax to array-like

File: config/common.php (hunk 8 above)
Comment (xepozz) [reply config/common.php]: Fixed

**Review by samdark — APPROVED** [review]

**Pull request merged**